### PR TITLE
[REVIEW] Improve Bsplines functions performance

### LINKED ISSUE #231: [FEA] Reevaluate Bsplines module for performance improvements
The Bsplines module need to be reevaluated to find areas of further performance improvements.

### PR DESCRIPTION
**Pull request by z-ryan1**
Closes #231 

This PR improves performance for all bsplines functions.
 
**Gauss Spline: branch-0.16**

```
---------- benchmark 'GaussSpline': 2 tests ---------
Name (time in us)                      Mean          
-----------------------------------------------------
test_gauss_spline_gpu[1-65536]     122.4854 (1.0)    
test_gauss_spline_cpu[1-65536]     714.0677 (5.83)   
-----------------------------------------------------
```

**Gauss Spline: This PR**
```
---------- benchmark 'GaussSpline': 2 tests ---------
Name (time in us)                      Mean          
-----------------------------------------------------
test_gauss_spline_gpu[1-65536]      27.1569 (1.0)    
test_gauss_spline_cpu[1-65536]     719.0537 (26.48)  
-----------------------------------------------------
```

**Cubic: branch-0.16**

```
-------- benchmark 'Cubic': 2 tests --------ns (2/2)
Name (time in us)             Mean          
--------------------------------------------
test_cubic_cpu[65536]     263.9371 (1.0)    
test_cubic_gpu[65536]     424.6950 (1.61)   
--------------------------------------------
```

**Cubic: This PR**
```
-------- benchmark 'Cubic': 2 tests --------ns (2/2)
Name (time in us)             Mean          
--------------------------------------------
test_cubic_gpu[65536]      15.3626 (1.0)    
test_cubic_cpu[65536]     261.7434 (17.04)  
--------------------------------------------
```

**Quadratic: branch-0.16**

```
--------- benchmark 'Quadratic': 2 tests ---------2)
Name (time in us)                   Mean          
--------------------------------------------------
test_quadratic_gpu[65536]       579.1895 (1.0)    
test_quadratic_cpu[65536]     1,140.8070 (1.97)   
--------------------------------------------------
```

**Quadratic: This PR**
```
--------- benchmark 'Quadratic': 2 tests ---------2)
Name (time in us)                   Mean          
--------------------------------------------------
test_quadratic_gpu[65536]        17.2439 (1.0)    
test_quadratic_cpu[65536]     1,149.6180 (66.67)  
--------------------------------------------------
```

The final function in `bsplines` is `cspline1d`.

**Cspline1d: branch-0.16**

```
--------- benchmark 'Cspline1d': 2 tests ---------2)
Name (time in ms)                   Mean          
--------------------------------------------------
test_cspline1d_cpu[65536]        46.0222 (1.0)    
test_cspline1d_gpu[65536]     2,564.3964 (55.72)  
--------------------------------------------------
```

After talking with Matt we found that the helper function `_quadratic_coeff` is dead code and should be removed. 

We also found that `cspline1d` uses both `_cubic_smooth_coeff` and `_cubic_coeff` which have data dependencies that should not be done on GPU version (see: line 187, 190, 202, 206 in [bsplines.py](https://github.com/rapidsai/cusignal/blob/ce096deddf61b527b54172adf3838d997f27c3b4/python/cusignal/bsplines/bsplines.py))

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by mnicely** [comment]
> After talking with Matt we found that the helper function `_quadratic_coeff` is dead code and should be removed.
> 
> We also found that `cspline1d` uses both `_cubic_smooth_coeff` and `_cubic_coeff` which have data dependencies that should not be done on GPU version (see: line 187, 190, 202, 206 in [bsplines.py](https://github.com/rapidsai/cusignal/blob/ce096deddf61b527b54172adf3838d997f27c3b4/python/cusignal/bsplines/bsplines.py))

@awthomp How do you feel about this?

**Comment by awthomp** [comment]
> > After talking with Matt we found that the helper function `_quadratic_coeff` is dead code and should be removed.
> > We also found that `cspline1d` uses both `_cubic_smooth_coeff` and `_cubic_coeff` which have data dependencies that should not be done on GPU version (see: line 187, 190, 202, 206 in [bsplines.py](https://github.com/rapidsai/cusignal/blob/ce096deddf61b527b54172adf3838d997f27c3b4/python/cusignal/bsplines/bsplines.py))
> 
> @awthomp How do you feel about this?

I'd keep it slow and on the GPU, within CuPy. If someone decides to use this function, ideally their E2E performance will make up for the sub-optimal GPU speed here.

**Comment by awthomp** [comment]
Spoke with @mnicely offline, and I took a look back at my original implementation notes. I have that `cspline1d` is a 'serial algorithm and revisit it'. I'm okay with removing it from the library, similarly to what we did with `lfilter`.

**Review by awthomp — APPROVED** [review]

**Comment by mnicely** [comment]
@z-ryan1 Please remove the dead and serial code.

File: python/cusignal/bsplines/bsplines.py [hunk 1]
```diff
@@ -32,57 +40,83 @@ def gauss_spline(x, n):
...
-    if cond2.any():
-        ax2 = ax[cond2]
-        res[cond2] = (ax2 - 1.5) ** 2 / 2.0
-    return res
+    x = cp.asarray(x)
+
+    return _quadratic_kernel(x)
 
 
 def _coeff_smooth(lam):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:104]: remove

File: python/cusignal/bsplines/bsplines.py [hunk 2]
```diff
@@ -32,57 +40,83 @@ def gauss_spline(x, n):
...
-    rho = (24 * lam - 1 - sqrt(xi)) / (24 * lam)
-    rho = rho * sqrt((48 * lam + 24 * lam * sqrt(3 + 144 * lam)) / xi)
+    xi = 1 - 96 * lam + 24 * lam * cp.sqrt(3 + 144 * lam)
+    omeg = cp.arctan2(cp.sqrt(144 * lam - 1), cp.sqrt(xi))
+    rho = (24 * lam - 1 - cp.sqrt(xi)) / (24 * lam)
+    rho = rho * cp.sqrt((48 * lam + 24 * lam * cp.sqrt(3 + 144 * lam)) / xi)
     return rho, omeg
 
 
 def _hc(k, cs, rho, omega):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:112]: remove

File: python/cusignal/bsplines/bsplines.py [hunk 3]
```diff
@@ -32,57 +40,83 @@ def gauss_spline(x, n):
...
+    return (
+        cs
+        / cp.sin(omega)
+        * (rho ** k)
+        * cp.sin(omega * (k + 1))
+        * cp.greater(k, -1)
+    )
 
 
 def _hs(k, cs, rho, omega):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:122]: remove

File: python/cusignal/bsplines/bsplines.py [hunk 4]
```diff
@@ -91,82 +125,84 @@ def _hs(k, cs, rho, omega):
...
+        / (1 - 2 * rho * rho * cp.cos(2 * omega) + rho ** 4)
     )
-    gamma = (1 - rho * rho) / (1 + rho * rho) / tan(omega)
+    gamma = (1 - rho * rho) / (1 + rho * rho) / cp.tan(omega)
     ak = abs(k)
-    return c0 * rho ** ak * (cos(omega * ak) + gamma * sin(omega * ak))
+    return c0 * rho ** ak * (cp.cos(omega * ak) + gamma * cp.sin(omega * ak))
 
 
 def _cubic_smooth_coeff(signal, lamb):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:135]: remove

File: python/cusignal/bsplines/bsplines.py [hunk 5]
```diff
@@ -91,82 +125,84 @@ def _hs(k, cs, rho, omega):
...
-            cs * yp[n] + 2 * rho * cos(omega) * y[n + 1] - rho * rho * y[n + 2]
+            cs * yp[n]
+            + 2 * rho * cp.cos(omega) * y[n + 1]
+            - rho * rho * y[n + 2]
         )
 
     return y
 
 
 def _cubic_coeff(signal):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:178]: remove

File: python/cusignal/bsplines/bsplines.py [hunk 6]
```diff
@@ -91,82 +125,84 @@ def _hs(k, cs, rho, omega):
...
-    powers = zi ** arange(K)
-    yplus[0] = signal[0] + zi * cp.sum(powers * signal)
-    for k in range(1, K):
-        yplus[k] = signal[k] + zi * yplus[k - 1]
-    output = zeros((K,), signal.dtype.char)
-    output[K - 1] = zi / (zi - 1) * yplus[K - 1]
-    for k in range(K - 2, -1, -1):
-        output[k] = zi * (output[k + 1] - yplus[k])
-    return output * 8.0
+# def _quadratic_coeff(signal):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:193]: remove

File: python/cusignal/bsplines/bsplines.py [hunk 7]
```diff
@@ -91,82 +125,84 @@ def _hs(k, cs, rho, omega):
...
+#     for k in range(1, K):
+#         yplus[k] = signal[k] + zi * yplus[k - 1]
+#     output = cp.zeros((K,), signal.dtype.char)
+#     output[K - 1] = zi / (zi - 1) * yplus[K - 1]
+#     for k in range(K - 2, -1, -1):
+#         output[k] = zi * (output[k + 1] - yplus[k])
+#     return output * 8.0
 
 
 def cspline1d(signal, lamb=0.0):
```
Comment (mnicely) [inline python/cusignal/bsplines/bsplines.py:208]: remove

**Review by mnicely — CHANGES_REQUESTED** [review]

**Comment by mnicely** [comment]
@awthomp @BradReesWork 
Do we wanna push this to 16 or 17?

**Review by mnicely — APPROVED** [review]

**Pull request merged**